File: android/src/main/kotlin/com/nexora/sdk/HardwareStorageManager.kt
package com.nexora.sdk

import android.content.Context
import android.os.StatFs
import java.io.File
import java.io.FileOutputStream

/**
 * Lightweight Storage Manager.
 * Provides safe file I/O, storage info, and directory access
 * without heavy external dependencies.
 */
class HardwareStorageManager(private val context: Context) {

    /// Returns total & available storage in bytes for internal and external storage.
    fun getStorageInfo(): Map<String, Any> {
        val internal = context.filesDir
        val external = context.getExternalFilesDir(null)
        
        val internalStat = StatFs(internal.absolutePath)
        val internalTotal = internalStat.blockSizeLong * internalStat.blockCountLong
        val internalFree = internalStat.blockSizeLong * internalStat.availableBlocksLong

        var externalTotal = 0L
        var externalFree = 0L
        if (external != null && external.exists()) {
            val externalStat = StatFs(external.absolutePath)
            externalTotal = externalStat.blockSizeLong * externalStat.blockCountLong
            externalFree = externalStat.blockSizeLong * externalStat.availableBlocksLong
        }

        return mapOf(
            "internalTotal" to internalTotal,
            "internalFree" to internalFree,
            "externalTotal" to externalTotal,
            "externalFree" to externalFree,
            "appCacheSize" to getDirSize(context.cacheDir),
            "appDataSize" to getDirSize(context.filesDir)
        )
    }

    /// Writes text content to a file inside app-private storage.
    fun writeFile(fileName: String, content: String): String {
        val file = safeFile(fileName)
        FileOutputStream(file).use { it.write(content.toByteArray()) }
        return file.absolutePath
    }

    /// Reads a text file from app-private storage.
    fun readFile(fileName: String): String? {
        val file = safeFile(fileName)
        if (!file.exists()) return null
        return file.readText()
    }

    /// Deletes a file from app-private storage.
    fun deleteFile(fileName: String): Boolean {
        val file = safeFile(fileName)
        return if (file.exists()) file.delete() else false
    }

    /// Checks whether a file exists in app-private storage.
    fun fileExists(fileName: String): Boolean {
        return safeFile(fileName).exists()
    }

    /// Lists all files in app-private storage directory.
    fun listFiles(): List<Map<String, Any>> {
        val dir = context.filesDir
        val files = dir.listFiles() ?: return emptyList()
        return files.map { f ->
            mapOf(
                "name" to f.name,
                "size" to f.length(),
                "isDirectory" to f.isDirectory,
                "lastModified" to f.lastModified()
            )
        }
    }

    /// Writes raw bytes to a file in app-private storage.
    fun writeBytes(fileName: String, bytes: ByteArray): String {
        val file = safeFile(fileName)
        FileOutputStream(file).use { it.write(bytes) }
        return file.absolutePath
    }

    /// Reads raw bytes from a file in app-private storage.
    fun readBytes(fileName: String): ByteArray? {
        val file = safeFile(fileName)
        if (!file.exists()) return null
        return file.readBytes()
    }

    /// Clears the app cache directory.
    fun clearCache(): Boolean {
        return try {
            deleteDir(context.cacheDir)
            true
        } catch (e: Exception) { false }
    }

    /// Returns the absolute path to the app-private files directory.
    fun getAppDirectory(): String = context.filesDir.absolutePath

    /// Returns the absolute path to the app-private cache directory.
    fun getCacheDirectory(): String = context.cacheDir.absolutePath

    /// Returns the external (shared) storage path if available.
    fun getExternalDirectory(): String? = context.getExternalFilesDir(null)?.absolutePath

    // --- Helpers ---

    private fun safeFile(fileName: String): File {
        require(isSafeFileName(fileName)) {
            "File name must be non-empty and cannot contain path separators."
        }
        val baseDir = context.filesDir.canonicalFile
        val file = File(baseDir, fileName).canonicalFile
        require(file.path.startsWith(baseDir.path + File.separator)) {
            "File path must stay inside app-private storage."
        }
        return file
    }

    private fun isSafeFileName(fileName: String): Boolean {
        return fileName.isNotBlank() &&
            fileName.length <= 120 &&
            !fileName.contains("/") &&
            !fileName.contains("\\") &&
            fileName != "." &&
            fileName != ".."
    }

    private fun getDirSize(dir: File): Long {
        if (!dir.exists()) return 0
        var size = 0L
        val files = dir.listFiles()
        if (files != null) {
            for (f in files) {
                size += if (f.isDirectory) getDirSize(f) else f.length()
            }
        }
        return size
    }

    private fun deleteDir(dir: File) {
        val files = dir.listFiles()
        if (files != null) {
            for (f in files) {
                if (f.isDirectory) deleteDir(f) else f.delete()
            }
        }
    }
}

File: ios/Classes/HardwareStorageManager.swift
import Foundation
import Flutter

/// Lightweight iOS Storage Manager.
/// Provides safe file I/O, storage diagnostics, and cache management.
public class HardwareStorageManager {

    /// Returns storage information for the device.
    public func getStorageInfo() -> [String: Any] {
        let fileManager = FileManager.default
        var internalTotal: Int64 = 0
        var internalFree: Int64 = 0

        if let attributes = try? fileManager.attributesOfFileSystem(forPath: NSHomeDirectory()) {
            internalTotal = (attributes[.systemSize] as? Int64) ?? 0
            internalFree = (attributes[.systemFreeSize] as? Int64) ?? 0
        }

        return [
            "internalTotal": internalTotal,
            "internalFree": internalFree,
            "externalTotal": 0,  // iOS has no external storage
            "externalFree": 0,
            "appCacheSize": getDirSize(getCacheDir()),
            "appDataSize": getDirSize(getAppDir())
        ]
    }

    /// Writes text content to a file in the Documents directory.
    public func writeFile(fileName: String, content: String) -> String? {
        guard let url = safeFileURL(fileName: fileName) else { return nil }
        do {
            try content.write(to: url, atomically: true, encoding: .utf8)
            return url.path
        } catch { return nil }
    }

    /// Appends text content to a file in the Documents directory.
    public func appendFile(fileName: String, content: String) -> String? {
        guard let url = safeFileURL(fileName: fileName) else { return nil }
        guard let data = content.data(using: .utf8) else { return nil }
        
        if FileManager.default.fileExists(atPath: url.path) {
            do {
                let fileHandle = try FileHandle(forWritingTo: url)
                defer { try? fileHandle.close() }
                try fileHandle.seekToEnd()
                try fileHandle.write(contentsOf: data)
                return url.path
            } catch {
                return nil
            }
        } else {
            do {
                try content.write(to: url, atomically: true, encoding: .utf8)
                return url.path
            } catch { return nil }
        }
    }

    /// Reads a text file from the Documents directory.
    public func readFile(fileName: String) -> String? {
        guard let url = safeFileURL(fileName: fileName) else { return nil }
        return try? String(contentsOf: url, encoding: .utf8)
    }

    /// Deletes a file from the Documents directory.
    public func deleteFile(fileName: String) -> Bool {
        guard let url = safeFileURL(fileName: fileName) else { return false }
        do {
            try FileManager.default.removeItem(at: url)
            return true
        } catch { return false }
    }

    /// Checks if a file exists in the Documents directory.
    public func fileExists(fileName: String) -> Bool {
        guard let url = safeFileURL(fileName: fileName) else { return false }
        return FileManager.default.fileExists(atPath: url.path)
    }

    /// Lists all files in the Documents directory.
    public func listFiles() -> [[String: Any]] {
        let dir = getAppDir()
        guard let contents = try? FileManager.default.contentsOfDirectory(at: dir, includingPropertiesForKeys: [.fileSizeKey, .isDirectoryKey, .contentModificationDateKey]) else {
            return []
        }
        return contents.map { url in
            let attrs = try? url.resourceValues(forKeys: [.fileSizeKey, .isDirectoryKey, .contentModificationDateKey])
            return [
                "name": url.lastPathComponent,
                "size": attrs?.fileSize ?? 0,
                "isDirectory": attrs?.isDirectory ?? false,
                "lastModified": Int64((attrs?.contentModificationDate?.timeIntervalSince1970 ?? 0) * 1000)
            ]
        }
    }

    /// Writes raw bytes to a file.
    public func writeBytes(fileName: String, bytes: FlutterStandardTypedData) -> String? {
        guard let url = safeFileURL(fileName: fileName) else { return nil }
        do {
            try bytes.data.write(to: url)
            return url.path
        } catch { return nil }
    }

    /// Reads raw bytes from a file.
    public func readBytes(fileName: String) -> FlutterStandardTypedData? {
        guard let url = safeFileURL(fileName: fileName) else { return nil }
        guard let data = try? Data(contentsOf: url) else { return nil }
        return FlutterStandardTypedData(bytes: data)
    }

    /// Clears the app cache directory.
    public func clearCache() -> Bool {
        let cacheDir = getCacheDir()
        guard let contents = try? FileManager.default.contentsOfDirectory(at: cacheDir, includingPropertiesForKeys: nil) else {
            return false
        }
        for url in contents {
            try? FileManager.default.removeItem(at: url)
        }
        return true
    }

    /// Returns the app Documents directory path.
    public func getAppDirectory() -> String {
        return getAppDir().path
    }

    /// Returns the app cache directory path.
    public func getCacheDirectory() -> String {
        return getCacheDir().path
    }

    /// Returns nil on iOS (no external storage).
    public func getExternalDirectory() -> String? {
        return nil
    }

    // MARK: - Private Helpers

    private func safeFileURL(fileName: String) -> URL? {
        guard isSafeFileName(fileName) else { return nil }
        let base = getAppDir().standardizedFileURL
        let url = base.appendingPathComponent(fileName, isDirectory: false).standardizedFileURL
        guard url.path.hasPrefix(base.path + "/") else { return nil }
        return url
    }

    private func isSafeFileName(_ fileName: String) -> Bool {
        return !fileName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty &&
            fileName.count <= 120 &&
            !fileName.contains("/") &&
            !fileName.contains("\\") &&
            fileName != "." &&
            fileName != ".."
    }

    private func getAppDir() -> URL {
        return FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
    }

    private func getCacheDir() -> URL {
        return FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0]
    }

    private func getDirSize(_ url: URL) -> Int64 {
        var size: Int64 = 0
        guard let enumerator = FileManager.default.enumerator(at: url, includingPropertiesForKeys: [.fileSizeKey]) else {
            return 0
        }
        for case let fileURL as URL in enumerator {
            let attrs = try? fileURL.resourceValues(forKeys: [.fileSizeKey])
            size += Int64(attrs?.fileSize ?? 0)
        }
        return size
    }
}
